enclave-runtime/src/lib.rs

```diff
@@ -383,57 +385,165 @@ pub unsafe extern "C" fn get_state(
 	sgx_status_t::SGX_SUCCESS
 }
 
-/// Call this once at worker startup to initialize the TOP pool and direct invocation RPC server.
+/// Initialize generic enclave components.
 ///
-/// This function will run the RPC server on the same thread as it is called and will loop there.
-/// That means that this function will not return as long as the RPC server is running. The calling
-/// code should therefore spawn a new thread when calling this function.
+/// Call this once at startup. Needs to be the first component initializer to be called.
 #[no_mangle]
-pub unsafe extern "C" fn init_direct_invocation_server(
-	server_addr: *const u8,
-	server_addr_size: usize,
-) -> sgx_status_t {
-	let mut server_addr_encoded = slice::from_raw_parts(server_addr, server_addr_size);
+pub unsafe extern "C" fn init_enclave_base_components() -> sgx_status_t {
+	let state_handler = Arc::new(GlobalFileStateHandler);
+	GLOBAL_STATE_HANDLER_COMPONENT.initialize(state_handler.clone());
 
-	let server_addr = match String::decode(&mut server_addr_encoded) {
+	let ocall_api = Arc::new(OcallApi);
+	GLOBAL_OCALL_API_COMPONENT.initialize(ocall_api.clone());
+
+	let stf_executor = Arc::new(EnclaveStfExecutor::new(ocall_api, state_handler));
+	GLOBAL_STF_EXECUTOR_COMPONENT.initialize(stf_executor);
+
+	sgx_status_t::SGX_SUCCESS
+}
+
+/// Initialize sidechain enclave components.
+///
+/// Call this once at startup. Has to be called AFTER the light-client
+/// (parentchain components) have been initialized (because we need the parentchain
+/// block import dispatcher).
+#[no_mangle]
+pub unsafe extern "C" fn init_enclave_sidechain_components() -> sgx_status_t {
+	let shielding_key = match Rsa3072Seal::unseal() {
+		Ok(k) => k,
+		Err(e) => {
+			error!("Failed to unseal shielding key: {:?}", e);
+			return sgx_status_t::SGX_ERROR_UNEXPECTED
+		},
+	};
+	let signer = match Ed25519Seal::unseal() {
 		Ok(s) => s,
 		Err(e) => {
-			error!("Decoding RPC server address failed. Error: {:?}", e);
+			error!("Error retrieving signer key pair: {:?}", e);
 			return sgx_status_t::SGX_ERROR_UNEXPECTED
 		},
 	};
-
-	let watch_extractor = Arc::new(create_determine_watch::<Hash>());
-	let connection_registry = Arc::new(ConnectionRegistry::<Hash, TungsteniteWsConnection>::new());
-
-	let rsa_shielding_key = match Rsa3072Seal::unseal() {
+	let state_key = match AesSeal::unseal() {
 		Ok(k) => k,
 		Err(e) => {
-			error!("Failed to unseal shielding key: {:?}", e);
+			error!("Failed to unseal state key: {:?}", e);
 			return sgx_status_t::SGX_ERROR_UNEXPECTED
 		},
 	};
 
-	let state_handler = Arc::new(GlobalFileStateHandler);
-	let ocall_api = Arc::new(OcallApi);
+	let stf_executor =
+		match GLOBAL_STF_EXECUTOR_COMPONENT.get() {
+			Some(a) => a,
+			None => {
+				error!("Failed to retrieve global STF executor component (maybe it is not initialized?)");
+				return sgx_status_t::SGX_ERROR_UNEXPECTED
+			},
+		};
```

Comment:
We could have some way more ergonomic code if we wrap this code into an internal function and introduce a custom error. Snippet:

```
fn init_enclave_sidechain_components_internal() -> EnclaveResult {
...
let state_handler = GLOBAL_STATE_HANDLER_COMPONENT
                          .get().ok_or_else(|| EnclaveError::Other("Could not get ...")?;
}
```

Have you considered this?

Reply:
Yes, I agree. I think I'll do a bit more refactoring to extract the initialization method into a separate module and move the initialization logic into there, with `Result` signatures 👍